zjit/src/codegen.rs

```diff
@@ -439,6 +439,24 @@ fn gen_get_ep(asm: &mut Assembler, level: u32) -> Opnd {
     ep_opnd
 }
 
+fn gen_objtostring(jit: &mut JITState, asm: &mut Assembler, val: Opnd, cd: *const rb_call_data, state: &FrameState) -> Option<Opnd> {
+    gen_prepare_non_leaf_call(jit, asm, state)?;
+
+    let iseq_opnd = Opnd::Value(jit.iseq.into());
+
+    // TODO: Specialize for immediate types
+    // Call rb_vm_objtostring(iseq, recv, cd)
+    let ret = asm_ccall!(asm, rb_vm_objtostring, iseq_opnd, val, (cd as usize).into());
+
+    // TODO: Call `to_s` on the receiver if rb_vm_objtostring returns Qundef
+    // Need to replicate what CALL_SIMPLE_METHOD does
+    asm_comment!(asm, "side-exit if rb_vm_objtostring returns Qundef");
+    asm.cmp(ret, Qundef.into());
+    asm.je(side_exit(jit, state, ObjToStringFallback)?);
```

Comment:
I'm not sure how to implement the conditional "call to_s on val when we got Qundef" without generating a lot of assembly (e.g. calling `gen_send_without_block` and `je` to it). So I decided to side exit in that case for now. This will at least fix the bug we're seeing
Thoughts?

Reply:
You can look at profiled objects, guard on the type (and side-exit), and generate only what's supposed to be executed. This is what YJIT does. It's generally useful to look at YJIT to build a baseline implementation for ZJIT.

Reply:
I'm not a fan of hiding a send in a non-send HIR instruction either due to optimizer visibility etc. One way around this is to have `insn::ObjToString` only call rb_vm_objtostring() and build the if-else logic into HIR during parsing. This way we have a regular send in the else BB, and the optimizer should still be able to handle folding away string input situations.

It still sounds quite involved to me though. I think side exiting for now is fine and we can make an issue and optimize this later.

Reply:
Thank you for your suggestions. I created https://github.com/Shopify/ruby/issues/701 as a follow up.

For now I'll keep side exit so we can avoid the incorrect behaviour.